rundeckapp/grails-app/services/rundeck/services/ProjectService.groovy

```diff
@@ -1721,13 +1717,61 @@ class ProjectService implements InitializingBean, ExecutionFileProducer, EventPu
      */
     @GrailsCompileStatic
     DeleteResponse deleteProject(IRundeckProject project, IFramework framework, AuthContext authContext, String username) {
-        def result = new DeleteResponse(success: false)
+        log.info("Requested deletion of project ${project.name}")
+
+        if(framework.getFrameworkProjectMgr().isFrameworkProjectDisabled(project.name)) {
+            return new DeleteResponse(
+            success: false,
+            error: "Cannot delete an already disabled project.")
+        }
+
+        if(!configurationService.getBoolean("projectService.deferredProjectDelete", false)) {
+            return deleteProjectInternal(project, framework, authContext, username)
+        }
+
+        try {
+            Project.withNewTransaction {
+                framework.getFrameworkProjectMgr().disableFrameworkProject(project.name)
+            }
+        } catch (UnsupportedOperationException e) {
+            log.warn("Could not disable project. Aborting project delete operation: ${e.getMessage()}", e)
+            throw new UnsupportedOperationException("Could not delete project. Project disabling not supported by framework: ${e.getMessage()}", e)
+        }
+
+        log.info("Deferring deletion of project ${project.name} to background task.")
+
+        def promise = Promises.task {
+            return Project.withNewTransaction {
+                return deleteProjectInternal(project, framework, authContext, username)
+            }
+        }
+        promise.onComplete { DeleteResponse it ->
+            log.info("Deletion of Project [${project.name}] finished with success [${it.success}]: ${it.error}")
+        }
+        promise.onError { Throwable t ->
+            log.error("Error deleting project [${project.name}]: ${t.getMessage()}", t)
+        }
+
+        return new DeleteResponse(success: true)
+    }
+
+
+    /**
+     * Delete a project completely
+     * @param project framework project
+     * @param framework frameowkr
+     * @return map [success:true/false, error: (String errorMessage)]
+     */
+    @GrailsCompileStatic
+    protected DeleteResponse deleteProjectInternal(IRundeckProject project, IFramework framework, AuthContext authContext, String username) {
+        log.info("Starting deletion of project ${project.name} by username $username")
         notify('projectWillBeDeleted', project.name)
+        def result = new DeleteResponse(success: false)
 
         //disable scm
         scmService.removeAllPluginConfiguration(project.name)
 
-        ScheduledExecution.withTransaction { TransactionStatus status ->
+        Project.withTransaction { TransactionStatus status ->
```

Comment:
Would rather leave this as a reference to ScheduledExecution so that it can be refactored out of a domain class specific implementation later.

Reply:
Reverted